FEATURE: 116 execute dataprocessor on result

DanielSiepmann opened the pull request:
Data processors can now be configured to be executed for each document while indexing AND for each search result after searching, before providing results to template.

Resolves: #116

DanielSiepmann committed:
TASK: Refactor data processing

Use own service to handle data processing. Classes like indexer should
not know anything about the structure and how to process the data.

Also rename record to data, as we can process just any data in form of
an array.

Relates: #116

DanielSiepmann committed:
TASK: Add data processing to search result

Search results are now processed through data processing by search
service. The result will be a SearchResult model from our domain. Also
SearchResult will execute new queries, e.g. from paginate widget,
through SearchService to apply data processing again.

Remove duplicate stub code to trait, to keep own logic and code clean.

DanielSiepmann committed:
TASK: Add test cases for ResultItem Model

Check whether all methods work as expected.

E.g. we can retrieve data in all ways, but not change anything.

DanielSiepmann committed:
TASK :Add tests for SearchRequest

Make sure exceptions with helpful messages are thrown if one object is
missing when execute is called.

Also make sure the expected methods are called.

DanielSiepmann committed:
TASK: Add tests for search result model

DanielSiepmann committed:
TASK: Add test for applied data processing on search result

DanielSiepmann committed:
TASK: Update documentation for data processing

DanielSiepmann committed:
Merge remote-tracking branch 'origin/develop' into feature/116-execute-dataprocessor-on-result

Conflicts:
    Classes/Domain/Index/AbstractIndexer.php
    Documentation/source/configuration/indexing.rst
    Tests/Unit/Domain/Index/AbstractIndexerTest.php

justusmoroni commented on Tests/Unit/Domain/Model/ResultItemTest.php:
```diff
@@ -0,0 +1,113 @@
+
+        $subject = new ResultItem($originalData);
+        $this->assertSame(
+            'Some title',
```
Use `$originalData['title']` here.

DanielSiepmann replied:
👍

justusmoroni commented on Tests/Unit/Domain/Model/ResultItemTest.php:
```diff
@@ -0,0 +1,113 @@
+            'uid' => 10,
+            'title' => 'Some title',
+        ];
+        $expectedData = $originalData;
```
Can't you remove this  and just use `$originalData` in line 41?

DanielSiepmann replied:
I could, but just to make sure that `$originalData` is not received as reference an modified, I use a copy.
This way I can make sure that I receive exactly the same as I passed into the constructor.

justusmoroni commented on Tests/Unit/Domain/Model/ResultItemTest.php:
```diff
@@ -0,0 +1,113 @@
+            'uid' => 10,
+            'title' => 'Some title',
+        ];
+        $expectedData = $originalData;
```
Not used?

DanielSiepmann replied:
Yeah copy and paste... thanks.

justusmoroni commented on Tests/Unit/Domain/Model/ResultItemTest.php:
```diff
@@ -0,0 +1,113 @@
+            'uid' => 10,
+            'title' => 'Some title',
+        ];
+        $expectedData = $originalData;
```
Not used?

DanielSiepmann replied:
Yeah copy and paste... thanks.

justusmoroni commented on Tests/Unit/Domain/Model/ResultItemTest.php:
```diff
@@ -0,0 +1,113 @@
+            'uid' => 10,
+            'title' => 'Some title',
+        ];
+        $expectedData = $originalData;
```
Not used?

DanielSiepmann replied:
Yeah copy and paste... thanks.

justusmoroni commented on Tests/Unit/Domain/Model/SearchResultTest.php:
```diff
@@ -0,0 +1,100 @@
+
+        $this->assertCount(3, $resultItems);
+
+        $this->assertSame(10, $resultItems[0]['uid']);
```
Use `$data[0]['uid]` here?
And the same thing in the following 2 lines?

DanielSiepmann replied:
👍

DanielSiepmann committed:
TASK: Improve ResultItemTest

Do not repeat content, use variable which is also better to read.
And do not add unnecessary, unused, variables.

DanielSiepmann committed:
TASK: Make test more readable

Make sure everyone knows what we compare, do not add hardcoded
information.
